crates/bitcoind_rpc/src/bip158.rs

```diff
@@ -6,262 +6,182 @@
+                height: next_height,
+                hash: next_hash,
+            });
+
+            let mut block = None;
+            let filter =
+                BlockFilter::new(self.client.get_block_filter(&next_hash)?.filter.as_slice());
+            if filter
+                .match_any(&next_hash, self.spks.iter().map(ScriptBuf::as_ref))
+                .map_err(Error::Bip158)?
+            {
+                block = Some(self.client.get_block(&next_hash)?);
+            }
 
-        // note: to connect with the local chain we must guarantee that `self.blocks.first()`
-        // is also the point of agreement with `self.cp`.
-        Some(
-            CheckPoint::from_block_ids(self.blocks.iter().map(BlockId::from))
-                .expect("blocks must be in order"),
-        )
+            // Store the next header
+            self.header = Some(next_header);
+            // Update self.cp
+            self.cp = cp.clone();
+
+            Ok(Some(Event { cp, block }))
+        })()
+        .transpose()
     }
 }
 
-/// Errors that may occur during a compact filters sync.
+/// Error that may be thrown by [`FilterIter`].
 #[derive(Debug)]
 pub enum Error {
-    /// bitcoin bip158 error
-    Bip158(bip158::Error),
-    /// attempted to scan blocks without any script pubkeys
-    NoScripts,
-    /// `bitcoincore_rpc` error
+    /// RPC error
     Rpc(bitcoincore_rpc::Error),
+    /// `bitcoin::bip158` error
+    Bip158(bitcoin::bip158::Error),
+    /// Max reorg depth exceeded.
+    ReorgDepthExceeded,
+    /// Error converting an integer
+    TryFromInt(core::num::TryFromIntError),
 }
 
+impl core::fmt::Display for Error {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Rpc(e) => write!(f, "{e}"),
+            Self::Bip158(e) => write!(f, "{e}"),
+            Self::ReorgDepthExceeded => write!(f, "maximum reorg depth exceeded"),
+            Self::TryFromInt(e) => write!(f, "{e}"),
+        }
+    }
+}
+
+#[cfg(feature = "std")]
+impl std::error::Error for Error {}
+
 impl From<bitcoincore_rpc::Error> for Error {
     fn from(e: bitcoincore_rpc::Error) -> Self {
         Self::Rpc(e)
     }
 }
 
-impl fmt::Display for Error {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        match self {
-            Self::Bip158(e) => e.fmt(f),
-            Self::NoScripts => write!(f, "no script pubkeys were provided to match with"),
-            Self::Rpc(e) => e.fmt(f),
-        }
+impl From<core::num::TryFromIntError> for Error {
+    fn from(e: core::num::TryFromIntError) -> Self {
+        Self::TryFromInt(e)
     }
 }
 
-#[cfg(feature = "std")]
-impl std::error::Error for Error {}
+/// Whether the RPC error is a "not found" error (code: `-5`).
+fn is_not_found(e: &bitcoincore_rpc::Error) -> bool {
+    matches!(
+        e,
+        bitcoincore_rpc::Error::JsonRpc(bitcoincore_rpc::jsonrpc::Error::Rpc(e))
+        if e.code == -5
+    )
+}
```

Comment:
question: is this the same as -5 from [Bitcoin Core docs](https://github.com/bitcoin/bitcoin/blob/36e40417de3fcaa776fb7bb8ebf3a23db4266572/src/rpc/protocol.h#L42C5-L42C31) ? I wonder if we should use it as not found or use the same as `invalid_address_or_key` as used there.